### PR TITLE
Mock Code Review

### PR DESCRIPTION
**Pull request by obloomfield**

### CODE REVIEW

File: mock/src/components/REPLInput.tsx [hunk 1]
```diff
@@ -1,102 +1,234 @@
...
-    } else if (command.startsWith("load_file")) {
-      const filePath = command.split(" ")[1];
-      return loadFile(filePath);
-      //return { filePath } + "loaded successfully.";
-    } else if (command == "view") {
-      viewFile();
-      return 'Now viewing file: "' + loadedFileName + '".';
+  //Function to search file
+  function searchFile(column: string, value: string): string {
+    var rowDisp = document.getElementById("rowDisp");
```
Comment (marissa-kate) [inline mock/src/components/REPLInput.tsx:120]: It's a bit unclear what the document.getElementById's function is, and the documentation on this method did not help in understanding. I am also concerned as to whether you are checking if there is a loaded file before searching, as you do in the view function. Would this method return "search completed successfully" if there is nothing loaded?

File: mock/src/components/REPLInput.tsx [hunk 2]
```diff
@@ -1,102 +1,234 @@
...
+    return loadFile(filePath);
+  });
+
+  //Registering the view command
+  commandRegistry.registerCommand("view", () => {
+    return viewFile();
+  });
+
+  //Registering the search command, returning error if incorrect format for search 
+  commandRegistry.registerCommand("search", (args) => {
```
Comment (marissa-kate) [inline mock/src/components/REPLInput.tsx:92]: I am concerned with not allowing the user to search without a column identifier. If you could check for the length being 1, you could add a null/" " element to args in order to allow the user to just type search <value> and the program to handle adapting the query. When you integrate with Server, you will need to handle this capability likely.

File: mock/src/components/REPLInput.tsx [hunk 3]
```diff
@@ -1,102 +1,234 @@
...
-      //return { filePath } + "loaded successfully.";
-    } else if (command == "view") {
-      viewFile();
-      return 'Now viewing file: "' + loadedFileName + '".';
+  //Function to search file
+  function searchFile(column: string, value: string): string {
+    var rowDisp = document.getElementById("rowDisp");
+    let columnIndex: number;
+    if (isNaN(Number(column))) {
+      columnIndex = loadedFile[0].indexOf(column);
```
Comment (marissa-kate) [inline mock/src/components/REPLInput.tsx:123]: This method appears to be handling a lot of the back-end functionality. I would have created mock search results that mimic what the Server would give you to help imagine what integration will look like. For instance, the checking of column indices validity can be handled by a back-end process most likely. By simplifying this method to delegate the searching, I think that you could more clearly separate these functionalities and not do double the work of checking twice.

File: mock/src/components/REPLInput.tsx [hunk 4]
```diff
@@ -1,102 +1,234 @@
...
     } else {
       loadedFile = mockedDataMap.get(filePath);
       loadedFileName = filePath;
     }
 
+    //Returning error if the file is empty
     if (loadedFile.length == 0) {
-      alert('The file: "' + fileName + '" is empty!');
-      return false;
+      alert('Error: The file: "' + fileName + '" is empty!');
```
Comment (marissa-kate) [inline mock/src/components/REPLInput.tsx:191]: The use of alerts is a very effective manner of communicating invalid interactions with the user. Are there any cases where you might not want to use it or might want to? For example, if the user puts in an invalid command, an alert might be useful, but something like "element with id 'rowDisp' not found." might not mean anything to a front end user. Consistency in how the error messages are displayed would be a good rule of thumb.

File: mock/src/components/REPLInput.tsx [hunk 5]
```diff
@@ -1,102 +1,234 @@
...
+  (args: Array<string>): string | string[][];
+}
+
+//Where the commands will be stored
+class CommandRegistry {
+  //Mapping the name of the command to the function
+  private commands = new Map<string, REPLFunction>();
+
+  //Registering commands
+  registerCommand(name: string, func: REPLFunction) {
```
Comment (marissa-kate) [inline mock/src/components/REPLInput.tsx:34]: Though this might be out of the scope of the project, since you have chosen to make a function to register commands, it is good to consider what happens if the developer tries to register a command with the same string command as something already in the map.

File: mock/src/components/REPLInput.tsx [hunk 6]
```diff
@@ -1,102 +1,234 @@
...
   mockedDataMap.set("property", PROPERTY_DATA);
   mockedDataMap.set("prices", PRICES_DATA);
   mockedDataMap.set("neighborhood", NEIGHBORHOOD_DATA);
   mockedDataMap.set("empty", EMPTY);
 
+  const commandRegistry = useRef(new CommandRegistry()).current;
+
+  //CHecking for mode switching
+  commandRegistry.registerCommand("mode", (args) => {
+    if (args[0] === "brief") {
```
Comment (priyam-parekh) [inline mock/src/components/REPLInput.tsx:66]: We also used strings to allow for mode change functionality. However, if we were to develop the program again, we would most likely use either an enum or a boolean becuase it would make it easier for other developers to use our implementation. You might also want to do the same!

File: mock/src/components/REPLInput.tsx (hunk 2 above)
Comment (priyam-parekh) [reply mock/src/components/REPLInput.tsx:92]: I definitely agree with Marissa here! Individuals using your program may be expecting this functionality, especially the stakeholders for whom this program was designed for—definitely something worth looking into!

File: mock/src/components/REPLInput.tsx [hunk 7]
```diff
@@ -1,102 +1,234 @@
...
+      }
+    }
+
+    //Searching the rows in the loaded file for the search value
+    const matchingRows = loadedFile
+      .slice(1)
+      .filter((row) => row[columnIndex] === value);
+    if (matchingRows.length === 0) {
+      alert(
+        `Error: No rows found with value "${value}" in column "${column}".`
```
Comment (priyam-parekh) [inline mock/src/components/REPLInput.tsx:144]: Great error checking here! It's nice to see that you included the specific value and column header name so that the user knows the exact issue with their input.

File: mock/src/components/REPLInput.tsx
Comment (priyam-parekh) [inline mock/src/components/REPLInput.tsx:1]: I personally believe that you could have extracted some of the logic from the REPLInput.tsx file and placed it into different files for better organization and extensibility. Currently, while I am able to track the functions and understand the underlying logic, it would be nice if the table generation was possibly in another file such as REPLHistory. This would not only help with debugging but also when you're changing or adapting parts of your code!

File: mock/tests/e2e/App.spec.ts [hunk 8]
```diff
@@ -54,16 +71,744 @@ test("after I type into the input box, its text changes", async ({ page }) => {
...
+    return commandElements.some((element) =>
+      element.textContent.includes(commandText)
+    );
+  }, "Switched to verbose mode");
+
+  // Assert that the command is found in the command history
+  expect(commandExists).toBeTruthy();
+});
+
+test("i can change the mode to verbose mode, and then back to brief more", async ({
```
Comment (priyam-parekh) [inline mock/tests/e2e/App.spec.ts:113]: It's great that you tested for both mode changes—from brief to verbose and then back.

File: mock/tests/e2e/App.spec.ts
Comment (priyam-parekh) [inline mock/tests/e2e/App.spec.ts:1]: I was really impressed by how comprehensive your edge case testing was. Especially with respect to checking case-sensitivity, searching with no results, particular columns not being found, out of bounds column indices, etc.
